File: proj/apps/composite_gray/composite.py
"""composite.py: Image compositing
"""

import numpy
import sys; sys.path += ['../../compiler']
import util
try:
    import numexpr
    has_numexpr = True
except ImportError:
    has_numexpr = False             # For PyPy comparisons

def composite(background, foreground, foreground_alpha, output_img):
    """Composites foreground on background using given alpha using loops."""

    for r in range(background.shape[0]):
        for c in range(background.shape[1]):
            alpha = foreground_alpha[r,c]
            output_img[r,c] = (1-alpha)*background[r,c] + alpha*foreground[r,c]

    return output_img

def composite_numpy(background, foreground, foreground_alpha):
    """Composites foreground on background using given alpha using numpy."""

    if foreground_alpha.shape != foreground.shape:
        foreground_alpha = numpy.dstack((foreground_alpha,)*foreground.shape[2])

    return (1-foreground_alpha)*background + foreground_alpha*foreground

def composite_numexpr(background, foreground, foreground_alpha):
    """Composites foreground on background using given alpha using numpy."""

    if foreground_alpha.shape != foreground.shape:
        foreground_alpha = numpy.dstack((foreground_alpha,)*foreground.shape[2])

    return numexpr.evaluate('(1-foreground_alpha)*background + foreground_alpha*foreground')


input_img_rgb = '../images/temple_rgb.png'
input_img_gray = '../images/temple_gray.png'
input_img2_rgb = '../images/house_rgb.png'
input_img2_gray = '../images/house_gray.png'
input_matte = input_img_gray

color_args = (input_img_rgb, input_img2_rgb, input_matte)
gray_args = (input_img_gray, input_img2_gray, input_matte)

def test(n=None):
    """
    Unit test for performance optimization.
    """
    
    L = []

    #if do_rgb:
    if False:
        L.append(util.test_image_pipeline_filename(composite, color_args, n, name='composite rgb', use_output_img=True))

    #if do_gray:
    if True:
        L.append(util.test_image_pipeline_filename(composite, gray_args, n, name='composite gray', use_output_img=True))
    
    return util.combine_tests(L)

def run_test_all(n=None):
    """
    Unit tests which compare various optimized programs with the original Python program output.
    """
    assert has_numexpr
    numexpr.set_num_threads(1)

    return util.combine_tests([
    util.test_image_pipeline_filename(composite, color_args, n, name='composite rgb'),
    util.test_image_pipeline_filename(composite, gray_args, n, grayscale=True, name='composite gray'),
    util.test_image_pipeline_filename(composite_numpy, color_args, n, name='composite_numpy rgb'),
    util.test_image_pipeline_filename(composite_numpy, gray_args, n, grayscale=True, name='composite_numpy gray'),
    util.test_image_pipeline_filename(composite_numexpr, color_args, n, name='composite_numexpr (1 thread) rgb'),
    util.test_image_pipeline_filename(composite_numexpr, gray_args, n, grayscale=True, name='composite_numexpr (1 thread) gray')])

if __name__ == '__main__':
    test()

File: proj/apps/composite/c/composite.c
#include "c_util.h"
#include <stdio.h>
#include <omp.h>

void composite(FLOAT3_IMG *background, FLOAT3_IMG *foreground, FLOAT_IMG *foreground_alpha, FLOAT3_IMG *output_img)
{
    int r, c;
    float alpha;
    
    #pragma omp parallel for private(c, alpha)
    for(r = 0; r < background->rows; r++)
    {
	for(c = 0; c < background->cols; c++)
	{
	    alpha = foreground_alpha->data[r][c];
	    output_img->data[r][c][0] = (1.0f - alpha) * background->data[r][c][0] + alpha * foreground->data[r][c][0];
	    output_img->data[r][c][1] = (1.0f - alpha) * background->data[r][c][1] + alpha * foreground->data[r][c][1];
	    output_img->data[r][c][2] = (1.0f - alpha) * background->data[r][c][2] + alpha * foreground->data[r][c][2];
	}
    }
}

int main(int argc, char **argv)
{
    FLOAT3_IMG *background;
    FLOAT3_IMG *foreground;
    FLOAT_IMG *foreground_alpha;
    
    if(argc == 1)
    {
	background = read_img_rgb("../../images/temple_rgb.png");
	foreground = read_img_rgb("../../images/house_rgb.png");
	foreground_alpha = read_img_gray("../../images/temple_gray.png");
    }
    else
    {
	if(argc == 5)
	{
	    background = read_img_rgb(argv[1]);
	    foreground = read_img_rgb(argv[2]);
	    foreground_alpha = read_img_gray(argv[3]);
	}
	else
	{
	    printf("Usage: ./a.out background.png foreground.png foreground_alpha.png out.png or ./a.out");
	    return 0;
	}
    }
    
    FLOAT3_IMG *output_img = zeros_rgb(background->rows, background->cols);
    
    double d1 = wall_time();
    composite(background, foreground, foreground_alpha, output_img);
    double d2 = wall_time();
    
    printf("%f\n", d2 - d1);
    
    if(argc == 1)
	write_img_rgb(output_img, "test.png");
    else
	write_img_rgb(output_img, argv[4]);
    
    return 0;
}
